apollo-router/src/compute_job/mod.rs

```diff
@@ -111,34 +132,46 @@ where
     F: FnOnce() -> T + Send + UnwindSafe + 'static,
     T: Send + 'static,
 {
-    let job_watcher = JobWatcher::new(compute_job_type);
-
-    let (tx, rx) = oneshot::channel();
-    let job = Box::new(move || {
-        // Ignore the error if the oneshot receiver was dropped
-        let _ = tx.send(std::panic::catch_unwind(job));
-    });
-
-    let job = Job {
-        ty: compute_job_type,
-        job_fn: job,
-        queue_start: Instant::now(),
-    };
-    queue().send(compute_job_type.into(), job);
-    async move {
-        let result = rx.await;
-        // This local variable MUST exist
-        let mut local_job_watcher = job_watcher;
-        local_job_watcher.outcome = match &result {
-            Ok(Ok(_)) => Outcome::ExecutedOk,
-            Ok(Err(_)) => Outcome::ExecutedError,
-            Err(_) => Outcome::ChannelError,
+    let compute_job_type_str: &'static str = compute_job_type.into();
+    let span = info_span!(
+        COMPUTE_JOB_SPAN_NAME,
+        "job.type" = compute_job_type_str,
+        "job.outcome" = tracing::field::Empty
+    );
+    span.in_scope(|| {
+        let job_watcher = JobWatcher::new(compute_job_type);
+
+        let (tx, rx) = oneshot::channel();
+        let job = Box::new(move || {
+            // Ignore the error if the oneshot receiver was dropped
+            let _ = tx.send(std::panic::catch_unwind(job));
+        });
+
+        let job = Job {
+            subscriber: Dispatch::default(),
+            parent_span: Span::current(),
+            ty: compute_job_type,
+            job_fn: job,
+            queue_start: Instant::now(),
         };
-        match result {
-            Ok(r) => r,
-            Err(e) => Err(Box::new(e) as Box<dyn Any + Send>),
+        queue().send(compute_job_type.into(), job);
+
+        async move {
+            let result = rx.await;
+            // This local variable MUST exist
```

Comment:
Why must it exist?

Reply:
Ahh, I should have put a better comment on this.

It's because without the local variable Rust will do a partial move of the job watcher causing it to drop before you expect it to.

It's actually part of Rust: https://github.com/rust-lang/rust/issues/108808